### ISSUE TITLE
WriteTagCheck: first top-level method's Javadoc tags not reported in compact source files (JEP 512)

### ISSUE THREAD

**Issue by vivek-0509** [checkstyle/checkstyle#20555]
Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/javadoc/writetag.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="tag" value="@since"/>
      <property name="tagFormat" value="\S+"/>
      <property name="tokens" value="METHOD_DEF"/>
      <property name="tagSeverity" value="error"/>
    </module>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputWriteTagCompact.java <<'EOF'
/*
 * File header block comment.
 */

/**
 * First method with @since.
 * @since 1.0
 */
void first() { }

/**
 * Second method also with @since.
 * @since 2.0
 */
void second() { }

void main() { }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputWriteTagCompact.java

Starting audit...
[ERROR] /Users/vivek/checkstyle/InputWriteTagCompact.java:12: Javadoc tag @since=. [WriteTag]
[ERROR] /Users/vivek/checkstyle/InputWriteTagCompact.java:13: Javadoc tag @since=2.0 [WriteTag]
Audit done.
Checkstyle ends with 2 errors.
```

---

**Describe what you expect in detail.**

WriteTagCheck must report the configured tag on every top-level method in a compact source file that has a Javadoc containing that tag. Given a compact file with two top-level methods, each with a `/** ... @since X */` Javadoc, the check should report the tag for both methods (four `[WriteTag]` messages: two per method's `@since` matches, same as the equivalent code inside an ordinary class).

Currently only the **second** method's Javadoc is scanned. The first method's Javadoc, the correctly-placed `/** ... @since 1.0 */` immediately preceding it, is silently ignored, so its two expected `@since` reports are missing from the output.

Root cause: when the file begins with a leading block comment (any `/* ... */` at the top of the file, whether it is a copyright header, an editor comment, or the xdocs `/*xml ... */` inline-config header), that comment is attached in the AST as a leading comment under the first top-level `METHOD_DEF`'s `TYPE` node in a compact source file. WriteTag reads the first attached leading block comment as the method's Javadoc, sees the header (which has no `@since` tags), and skips further Javadoc scanning for that method. The actual `/** */` Javadoc is bypassed. In an ordinary class file the same header attaches to the outer `CLASS_DEF` instead, so the first method's `/** */` block is correctly identified as its Javadoc and its tags are reported.

Expected on the compact input: four violations, matching the ordinary-class output shown below.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputWriteTagOrdinary.java <<'EOF'
/*
 * File header block comment.
 */

public class InputWriteTagOrdinary {

    /**
     * First method with @since.
     * @since 1.0
     */
    public void first() { }

    /**
     * Second method also with @since.
     * @since 2.0
     */
    public void second() { }

    public void main() { }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -c config.xml InputWriteTagOrdinary.java

Starting audit...
[ERROR] /Users/vivek/checkstyle/InputWriteTagOrdinary.java:8: Javadoc tag @since=. [WriteTag]
[ERROR] /Users/vivek/checkstyle/InputWriteTagOrdinary.java:9: Javadoc tag @since=1.0 [WriteTag]
[ERROR] /Users/vivek/checkstyle/InputWriteTagOrdinary.java:14: Javadoc tag @since=. [WriteTag]
[ERROR] /Users/vivek/checkstyle/InputWriteTagOrdinary.java:15: Javadoc tag @since=2.0 [WriteTag]
Audit done.
Checkstyle ends with 4 errors.
```

**Comment by romani** [comment]
better to not start fixes until https://github.com/checkstyle/checkstyle/issues/19150 is done, such PR will change implementation of Check significantly.

**Comment by ZaheerAhmadDev** [comment]
After the merge of https://github.com/checkstyle/checkstyle/issues/19150:
Looks like we have three problems:

```
ZaheerAhmad@fedora:~/checks$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WriteTag">
            <property name="tag" value="@since"/>
            <property name="tagFormat" value="\S+"/>
            <property name="tokens" value="METHOD_DEF"/>
            <property name="tagSeverity" value="error"/>
        </module>
    </module>
</module>
```

```java
ZaheerAhmad@fedora:~/checks$ cat src/Test.java
/*
 * File header block comment.
 */

/**
 * First method with @since.
 * @since 1.0
 */
void first() { }

/**
 * Second method also with @since.
 * @since 2.0
 */
void second() { }

void main() { }
```

```
ZaheerAhmad@fedora:~/checks$ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] /home/ZaheerAhmad/checks/src/Test.java:13: Javadoc tag @since=2.0 [WriteTag]
Audit done.
Checkstyle ends with 1 errors.
ZaheerAhmad@fedora:~/checks$ 
```
it should show violation on compact source.

same inside ordinary class it is not showing violation `.`.

```
ZaheerAhmad@fedora:~/checks$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="WriteTag">
            <property name="tag" value="@since"/>
            <property name="tagFormat" value="\S+"/>
            <property name="tokens" value="METHOD_DEF"/>
            <property name="tagSeverity" value="error"/>
        </module>
    </module>
</module>
```

```java
ZaheerAhmad@fedora:~/checks$ cat src/Test.java
/*
 * File header block comment.
 */

public class Test {

    /**
     * First method with @since.
     * @since 1.0
     */
    public void first() { }

    /**
     * Second method also with @since.
     * @since 2.0
     */
    public void second() { }

    public void main() { }
}
```

```
ZaheerAhmad@fedora:~/checks$ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] /home/ZaheerAhmad/checks/src/Test.java:9: Javadoc tag @since=1.0 [WriteTag]
[ERROR] /home/ZaheerAhmad/checks/src/Test.java:15: Javadoc tag @since=2.0 [WriteTag]
Audit done.
Checkstyle ends with 2 errors.
ZaheerAhmad@fedora:~/checks$ 
```
Question:
Is the violation removed on line `8` and line `14`  intentionaly with updation of check?

From https://github.com/checkstyle/checkstyle/issues/21200
check is also showing `Caused by: java.lang.AssertionError: Misuse of notNull: argument was null`

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="tag" value="@since"/>
      <property name="tokens" value="ENUM_CONSTANT_DEF"/>
    </module>
  </module>
</module>
```
```java
ZaheerAhmad@fedora:~/checks$ cat src/Test.java
public enum Test {

    /**
     * doc for FIRST
     */
    FIRST,

    SECOND,

    /**
     * doc for THIRD
     */
    @Deprecated
    THIRD
}
```

```
ZaheerAhmad@fedora:~/checks$ java $RUN_LOCALE -jar checkstyle-13.10.0-all.jar -c config.xml src/Test.java
Starting audit...
Exception in thread "main" java.lang.Error: Error was thrown while processing src/Test.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:321)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:227)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:429)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:347)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:205)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:129)
Caused by: java.lang.AssertionError: Misuse of notNull: argument was null
        at com.puppycrawl.tools.checkstyle.utils.NullUtil.notNull(NullUtil.java:54)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.findJavadoc(WriteTagCheck.java:278)
        at com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck.visitToken(WriteTagCheck.java:191)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:384)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:457)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:322)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:206)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:108)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:341)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:300)
        ... 5 more
ZaheerAhmad@fedora:~/checks$ 
```
i checked locally the root cause is 


<img width="1174" height="229" alt="Image" src="https://github.com/user-attachments/assets/208d7432-0a6b-43ef-88a2-1707eaa07d9d" />


https://github.com/checkstyle/checkstyle/blob/6a0e8f47a073e2bcc85a50d47570ffa6b8001469/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheck.java#L250

we need to remove `NullUtil.notNull` and would need to add a condition if modifier is not null something like this:
```java
    private static DetailAST findJavadoc(DetailAST ast) {
        DetailAST cmt = ast.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
        if (cmt == null) {
            final DetailAST modifiers = ast.findFirstToken(TokenTypes.MODIFIERS);
            final DetailAST type = ast.findFirstToken(TokenTypes.TYPE);
            if( modifiers != null ) {
                final DetailAST annotation = modifiers.findFirstToken(TokenTypes.ANNOTATION);

                cmt = modifiers.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
                if (annotation != null) {
                    cmt = annotation.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
                }
            }
            if (cmt == null && type != null) {
                cmt = type.findFirstToken(TokenTypes.BLOCK_COMMENT_BEGIN);
            }
        }
        return cmt;
    }
```
> [ERROR] /Users/vivek/checkstyle/InputWriteTagOrdinary.java:8: Javadoc tag @since=. [WriteTag]

@romani after the merge of https://github.com/checkstyle/checkstyle/pull/19193, `is the Javadoc tag @since=.`  intentionally removed?

**Comment by romani** [comment]
>Question:
Is the violation removed on line 8 and line 14 intentionaly with updation of check?

Yes. Check got heavy update to do validation on missing tag ONLY 
https://github.com/checkstyle/checkstyle/issues/21080
